### PR TITLE
build.js: expand where bindings regex to include '[' and ']'

### PR DESCRIPTION
**Pull request by xwvvvvwx**
Some matches in storage writes were being missed as a result of this oversight.

### CODE REVIEW

File: examples/where/src/lemmas.k.md [hunk 1]
```diff
@@ -1,5 +1,8 @@
 ```k
 rule A -Word B <=Int A => #rangeUInt(256, A -Int B)
   requires #rangeUInt(256, A) andBool #rangeUInt(256, B)
+
+rule A <=Int chop(A +Int B) => #rangeUInt(256, A +Int B)
+  requires #rangeUInt(256, A) andBool #rangeUInt(256, B)
```
Comment (asymmetric) [inline examples/where/src/lemmas.k.md]: Is the `requires` really required? Seems like it might be redundant.

Comment (asymmetric) [reply examples/where/src/lemmas.k.md]: Also, when does this rule help?

File: examples/where/dapp/src/Calculate.sol [hunk 2]
```diff
@@ -11,9 +11,12 @@ contract Calculate {
         require(y == 0 || (z = x * y) / y == x);
     }
 
-    function run(uint a, uint b) public pure returns (uint e) {
+    uint result = 0;
```
Comment (asymmetric) [inline examples/where/dapp/src/Calculate.sol]: Sanity check (mine): this would be 0 even without the assignment, right?

File: lib/build.js [hunk 3]
```diff
@@ -455,19 +455,21 @@ const parseAct = config => (act_str, silent = false) => {
...
+                 *  - ']'
                  *  - the end of a line
                  *  - a tab
                  *  - a space
                  *
-                 *  example here: https://regex101.com/r/ZubkN1/3
+                 *  example here: https://regex101.com/r/ZubkN1/4
                  */
-                "from": new RegExp(`(, ?|\\(|^|\\t| )(${l.split(":=")[0].trim()})(,|\\)|$|\\t| )`, "gm"),
+                "from": new RegExp(`(, ?|\\(|\\[|^|\\t| )(${l.split(":=")[0].trim()})(,|\\)|$|\\t|\\]| )`, "gm"),
```
Comment (asymmetric) [inline lib/build.js]: Could you explain this regex to me?

This is matching the LHS of a `where` assignment, right? Like

```act
where
  A := B
```

When does the LHS contain brackets etc? I feel like I'm missing something.

File: examples/where/dapp/src/Calculate.sol [hunk 4]
```diff
@@ -11,9 +11,12 @@ contract Calculate {
...
+    function run(uint a, uint b) public returns (uint) {
         uint c = sub(a, b);
         uint d = mul(c, b);
-        e = add(d, a);
+        uint e = add(d, a);
+
+        result = e; return e;
```
Comment (asymmetric) [inline examples/where/dapp/src/Calculate.sol]: I would split these but I also don't care. ;)

File: lib/build.js (hunk 3 above)
Comment (xwvvvvwx) [reply lib/build.js]: the brackets are capture groups. It's the only way I could find to make the regex work nice. I would be happy if you know how to do it without.

File: examples/where/dapp/src/Calculate.sol (hunk 2 above)
Comment (xwvvvvwx) [reply examples/where/dapp/src/Calculate.sol]: I believe that is the case yes, but I think it's normally good to be explicit.

**Comment by xwvvvvwx** [comment]
> Can you give an example of when this was failing? Maybe even adding them to the actual example, so that we're sure it won't fail again.

Apologies, should have been more clear in the commit message. The regex was failing to match when the LHS was present in the rightmost position of a storage rewrite:

e.g.

```
[#Contract.foo[Bar] <- FooBar => Baz]
```

Where `Baz` is the value that should be replaced by the regex. I found it while playing around with [this spec](https://github.com/dapphub/klab/blob/concrete_maps/examples/concrete_maps/src/specification.act.md), where the `BURN` in the write to `balanceOf[0]` was not being replaced correctly.

The example has already been updated to cover this case.

File: examples/where/src/lemmas.k.md (hunk 1 above)
Comment (xwvvvvwx) [reply examples/where/src/lemmas.k.md]: The rule is needed to help `K` deal with the overflow checking introduced by solidity. The requires may not be needed, I have not tried without them. I'll try and run the spec without them and see what happens.

File: lib/build.js (hunk 3 above)
Comment (xwvvvvwx) [reply lib/build.js]: There is a fairly complete set of examples for the regex [here](Could you explain this regex to me?). This is documented in the comment above.

File: examples/where/src/lemmas.k.md (hunk 1 above)
Comment (xwvvvvwx) [reply examples/where/src/lemmas.k.md]: turns out they were not required. removed. thx!

File: lib/build.js (hunk 3 above)
Comment (asymmetric) [inline lib/build.js]: Why do you need to doube escape with `\\[`? The regex in the link doesn't need to do that for some reason.

Comment (mhhf) [reply lib/build.js]: In the link, a regex is constructed with `/<regex>/`
here an object of "class" RegExp is constructed and a String passed to its constructor
in a String the `\` character is an escape character used for place strange characters in a string: e.g. `\t` is a tab `\n` new line, etc. so in order to type the '\' character in a string as exactly that character you need to escape it: `\\`
the final string object will only contain `\[` in this case

**Pull request merged**